### PR TITLE
Fix software to use fetch/find

### PR DESCRIPTION
**Pull request by kress95**
#### Planned changes:

- [x] Update tests to validate struct type returns.
- [x] Fix StorageDrive API (I still didn't get the best way to do it, I'm really up for suggestions)

---
This change is [Reviewable](https://reviewable.io/reviews/hackerexperience/helix/6)

### CODE REVIEW

**Comment by mememori** [comment]
If your PR is a WIP, add a WIP tag

File: apps/software/lib/software/controller/file_module.ex [hunk 1]
```diff
@@ -29,11 +31,11 @@ defmodule Helix.Software.Controller.FileModule do
...
   end
 
-  @spec find(File.t) :: module_roles
-  def find(file) do
+  @spec get_file_modules(File.t | File.id) :: module_roles
+  def get_file_modules(file) do
     file
     |> FileModule.Query.from_file()
-    |> FileModule.Query.select_module_role_id_and_module_version()
+    |> select([m], {m.module_role_id, m.module_version})
```
Comment (kress95) [inline apps/software/lib/software/controller/file_module.ex]: change what exactally? the field name on db?
I'm up for suggestions :smiley:

Comment (mememori) [inline apps/software/lib/software/controller/file_module.ex]: the table/record/schema name (`ModuleRole` and `FileModule`) and the field name (used interchangeably, they are `role`, `module` and `module_role`)

Comment (mememori) [inline apps/software/lib/software/controller/file_module.ex]: obviously not on this PR

File: apps/software/lib/software/controller/storage_drive.ex [hunk 2]
```diff
@@ -1,33 +1,37 @@
...
+    do: StorageDrive.Query.from_storage(query, storage)
+
+  @spec delete(Storage.t | HELL.PK.t, integer) :: no_return
+  def delete(storage, drive_id) do
+    storage
+    |> StorageDrive.Query.from_storage()
+    |> StorageDrive.Query.by_drive_id(drive_id)
     |> Repo.delete_all()
 
     :ok
```
Comment (kress95) [inline apps/software/lib/software/controller/storage_drive.ex]: I still didn't get the best way to do it, so I'm skipping that until receiving suggestions.

**Comment by kress95** [comment]
I'm fixing #8 and #14 here

**Comment by mememori** [comment]
> I'm fixing #8 and #14 here

I was thinking about considering letting you do that to avoid requiring you to mess with your work tree. But thinking about it better, i'll simply disallow that.

Neither issues depend on this PR and you piping them here will just make my life harder and make this PR prone to pass with bad code that wasn't noticed thanks to the sheer amount of LoC changes

File: apps/software/lib/software/controller/storage_drive.ex (hunk 2 above)
Comment (mememori) [inline apps/software/lib/software/controller/storage_drive.ex]:
```elixir
get_storage_drives(Storage.t) :: [PK.t]
link_drive(Storage.t, PK.t) :: :ok | {:error, reason :: term}
unlink_drive(PK.t) :: :ok
```

anything else is a waste of time.

Comment (mememori) [inline apps/software/lib/software/controller/storage_drive.ex]: Also, #14 derives from this branch (but in a separate PR) for this very reason

**Comment by kress95** [comment]
I believe this is ready for revision, any reason (but the WIP tag) to remove it from there, @mememori ?

**Comment by mememori** [comment]
:lgtm:

---

Reviewed 8 of 11 files at r1, 3 of 3 files at r2, 16 of 23 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/helix/6#-:-KfrzneCzuhMlbFHV5rL:bnfp4nl)*

**Comment by mememori** [comment]
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks broke.

---

*[apps/helix_software/lib/software/controller/storage_drive.ex, line 21 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs4CuX2CgKBr6gBMIu:-Kfs4CuX2CgKBr6gBMIv:b-onagbn) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/lib/software/controller/storage_drive.ex#L21)):*
> ```Elixir
>       {:error, _} ->
>         {:error, :internal}
>     end
> ```

Put a TODO here to later check if the error changeset is telling that the UNIQ constraint is being violated (to return an error like `already_in_use`

---

*[apps/helix_software/test/controller/file_module_test.exs, line 71 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs3d7HZ5jqle0AwRrg:-Kfs3d7HZ5jqle0AwRrh:b-wi4ly3) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/file_module_test.exs#L71)):*
> ```Elixir
> 
>       assert {:error, :notfound} ==
>         FileModuleController.update(file, module_id, version)
> ```

Avoid breaking an assert in two lines, it looks weird. Try to bind the value and compare

---

*[apps/helix_software/test/controller/file_module_test.exs, line 85 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs3TkBA4umjdFbcSbr:-Kfs3TkBA4umjdFbcSbs:b-6mydua) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/file_module_test.exs#L85)):*
> ```Elixir
>     file_modules = FileModuleController.get_file_modules(file)
> 
>     refute Enum.empty?(module_roles)
> ```

This check is completely unnecessary, if you can't believe on your own generated input, then you can't test anything

---

*[apps/helix_software/test/controller/file_test.exs, line 77 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs3KKAP5nyNaADNC-M:-Kfs3KKAP5nyNaADNC-N:b-85p56i) ([raw file](https://github.com/hackerexperience/helix/blob/2d3f497c7be28932c3cda797664aabe7d2ac4fcd/apps/helix_software/test/controller/file_test.exs#L77)):*
> ```Elixir
>     # Storage
>     # REVIEW: I think we should disallow this. A file should not be
>     #   moved/updated to another storage but should be copied to it
> ```

Why did you remove this ? 

---

*[apps/helix_software/test/controller/file_test.exs, line 176 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs35K7oGbAZjbq3N1p:-Kfs35K7oGbAZjbq3N1q:b-lvr5ta) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/file_test.exs#L176)):*
> ```Elixir
>         |> Factory.build()
>         |> Map.merge(similarities)
>         |> Factory.insert()
> ```

`Factory.insert(:file, similarities)`

---

*[apps/helix_software/test/controller/file_test.exs, line 204 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs3-IUwPBBTICkuyNH:-Kfs3-IUwPBBTICkuyNI:bwt8tl0) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/file_test.exs#L204)):*
> ```Elixir
>       FileController.delete(file.file_id)
>       refute FileController.fetch(file.file_id)
>     end
> ```

Make those separate tests

---

*[apps/helix_software/test/controller/file_text_test.exs, line 17 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs2unAEWBMLW7F5cMV:-Kfs2unAEWBMLW7F5cMW:b-eagw7p) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/file_text_test.exs#L17)):*
> ```Elixir
> 
>     %{file_id: file_id, contents: contents}
>   end
> ```

Wrong, a file_text should create a file, not receive one

---

*[apps/helix_software/test/controller/file_text_test.exs, line 31 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs1XyXUH09XlLTlq0A:-Kfs1XyXUH09XlLTlq0B:b-xapvkh) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/file_text_test.exs#L31)):*
> ```Elixir
>       assert {:error, cs} = FileTextController.create(params)
>       assert :file_id in Keyword.keys(cs.errors)
>     end
> ```

This is wrong. `FileTextController` is an specialized file, so it should not receive a potential file and put contents on it but it should create a file to hold it's contents

---

*[apps/helix_software/test/controller/file_text_test.exs, line 44 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs1EUO4ZFyAbwMrS77:-Kfs1EUO4ZFyAbwMrS78:bk3x7pw) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/file_text_test.exs#L44)):*
> ```Elixir
>         :file
>         |> Factory.build()
>         |> Map.put(:file_id, PK.pk_for(File))
> ```

`Factory.build(:file, %{file_id: Pk.pk_for(File)})`

---

*[apps/helix_software/test/controller/storage_drive_test.exs, line 16 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs10UZMg7N2kxm4ay-:-Kfs10UZMg7N2kxm4ay0:bow7vns) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/storage_drive_test.exs#L16)):*
> ```Elixir
>     |> Map.put(:drives, [])
>     |> Factory.insert()
>   end
> ```

`Factory.insert(:storage, %{drives: []})`

---

*[apps/helix_software/test/controller/storage_drive_test.exs, line 37 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs3osvLFUOnzDu0Of2:-Kfs3osvLFUOnzDu0Of3:boi3op2) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/storage_drive_test.exs#L37)):*
> ```Elixir
> 
>     refute Enum.empty?(expected_drives)
>     assert Enum.empty?(expected_drives -- got_drives)
> ```

This looks awkward.

`assert Enum.sort(expected_drives) == Enum.sort(got_drives)`

---

*[apps/helix_software/test/controller/storage_drive_test.exs, line 42 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs3zOVq5ujHWBTengX:-Kfs3zOVq5ujHWBTengY:bh8i57q) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/storage_drive_test.exs#L42)):*
> ```Elixir
>     got_drives = Controller.get_storage_drives(driveless_storage)
> 
>     assert Enum.empty?(got_drives)
> ```

This should be in a different test

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/helix/6)*

**Comment by mememori** [comment]
Fix the above issues and rebase. I've promoted the FileText to an issue (with it's own PR) #34

**Comment by kress95** [comment]
> I thought we had discussed and decided that FileText is an implementation of File so it should create a File automatically, instead of receiving it as a second step that could happen at any time...

This really needs to be fixed on a PR about fixing `fetch/find`?

**Comment by kress95** [comment]
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks broke.

---

*[apps/helix_software/lib/software/controller/storage_drive.ex, line 21 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs4CuX2CgKBr6gBMIu:-Kfv3GIWs1Di0imiBulT:b-896fix) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/lib/software/controller/storage_drive.ex#L21)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

Put a TODO here to later check if the error changeset is telling that the UNIQ constraint is being violated (to return an error like `already_in_use`
</blockquote></details>

Done.

---

*[apps/helix_software/test/controller/file_module_test.exs, line 71 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs3d7HZ5jqle0AwRrg:-Kfv3HdU3jcBqPUOF9W7:b-896fix) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/file_module_test.exs#L71)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

Avoid breaking an assert in two lines, it looks weird. Try to bind the value and compare
</blockquote></details>

Done.

---

*[apps/helix_software/test/controller/file_module_test.exs, line 85 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs3TkBA4umjdFbcSbr:-Kfv3J2rlf77aRhLe7zM:b-896fix) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/file_module_test.exs#L85)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

This check is completely unnecessary, if you can't believe on your own generated input, then you can't test anything
</blockquote></details>

Done.

---

*[apps/helix_software/test/controller/file_test.exs, line 77 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs3KKAP5nyNaADNC-M:-Kfv3lnCBWd3LLTaUthw:b-896fix) ([raw file](https://github.com/hackerexperience/helix/blob/2d3f497c7be28932c3cda797664aabe7d2ac4fcd/apps/helix_software/test/controller/file_test.exs#L77)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

Why did you remove this ? 
</blockquote></details>

Done.

---

*[apps/helix_software/test/controller/file_test.exs, line 176 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs35K7oGbAZjbq3N1p:-Kfv4DCtwX9wUBHpLT1h:b-896fix) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/file_test.exs#L176)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

`Factory.insert(:file, similarities)`
</blockquote></details>

Done.

---

*[apps/helix_software/test/controller/file_test.exs, line 204 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs3-IUwPBBTICkuyNH:-Kfv4U1Uc2UzznDH4RCR:b-4qjmza) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/file_test.exs#L204)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

Make those separate tests
</blockquote></details>

I totally let this pass, sorry, if I had seen that I wouldn't even ask for revision, I really need to create a better self revision process.

Anyways, it's done.

---

*[apps/helix_software/test/controller/file_text_test.exs, line 17 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs2unAEWBMLW7F5cMV:-Kfv7Es39_DqBDzD59YD:bw7u5rc) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/file_text_test.exs#L17)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

Wrong, a file_text should create a file, not receive one
</blockquote></details>

I'll fix that on another revision

---

*[apps/helix_software/test/controller/file_text_test.exs, line 31 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs1XyXUH09XlLTlq0A:-Kfv7Hi_gDQIxBesO6bk:bw7u5rc) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/file_text_test.exs#L31)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

This is wrong. `FileTextController` is an specialized file, so it should not receive a potential file and put contents on it but it should create a file to hold it's contents
</blockquote></details>

I'll fix that on another revision

---

*[apps/helix_software/test/controller/file_text_test.exs, line 44 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs1EUO4ZFyAbwMrS77:-Kfv4szJozfUvAhUrv6N:b6i182g) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/file_text_test.exs#L44)):*
> Factory.build(:file, %{file_id: Pk.pk_for(File)})

Looks really better, thanks.

---

*[apps/helix_software/test/controller/storage_drive_test.exs, line 16 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs10UZMg7N2kxm4ay-:-Kfv5Y7FcYHfBd20cZG9:b-896fix) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/storage_drive_test.exs#L16)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

`Factory.insert(:storage, %{drives: []})`
</blockquote></details>

Done.

---

*[apps/helix_software/test/controller/storage_drive_test.exs, line 37 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs3osvLFUOnzDu0Of2:-Kfv5fb6dG5xhH9Pmpm1:b-896fix) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/storage_drive_test.exs#L37)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

This looks awkward.

`assert Enum.sort(expected_drives) == Enum.sort(got_drives)`
</blockquote></details>

Done.

---

*[apps/helix_software/test/controller/storage_drive_test.exs, line 42 at r3](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfs3zOVq5ujHWBTengX:-Kfv5s_FO9AVncxmsX6E:b-896fix) ([raw file](https://github.com/hackerexperience/helix/blob/5e407c983fc16f2e690637ad8569f6df12357af2/apps/helix_software/test/controller/storage_drive_test.exs#L42)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

This should be in a different test
</blockquote></details>

Done.

---

*[apps/software/lib/software/controller/storage_drive.ex, line 37 at r1](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfhah-M-P2WIjMjMW1Q-r1-37:-Kfv6Cyen9kcDMXHfJal:b44uned) ([raw file](https://github.com/hackerexperience/helix/blob/30c56407a77c72b5e5e24a8215c4e026d32cc3a1/apps/software/lib/software/controller/storage_drive.ex#L37)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

Also, #14 derives from this branch (but in a separate PR) for this very reason
</blockquote></details>

This API is implemented, I'll move it to storage and remove StorageDrive controller on a separated revision.

---

*[apps/software/test/controller/file_module_test.exs, line 86 at r1](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfhah-M-P2WIjMjMW1S-r1-86:-Kfv6ByVJ7cLqe9Sp70r:b-896fix) ([raw file](https://github.com/hackerexperience/helix/blob/30c56407a77c72b5e5e24a8215c4e026d32cc3a1/apps/software/test/controller/file_module_test.exs#L86)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

`Enum.empty?` is more descriptive
</blockquote></details>

Done.

---

*[apps/software/test/controller/file_test.exs, line 85 at r1](https://reviewable.io:443/reviews/hackerexperience/helix/6#-Kfhah-M-P2WIjMjMW1T-r1-85:-Kfv66-8WfPL7-jpbmVC:b-896fix) ([raw file](https://github.com/hackerexperience/helix/blob/30c56407a77c72b5e5e24a8215c4e026d32cc3a1/apps/software/test/controller/file_test.exs#L85)):*
<details><summary><i>Previously, mememori (Charlotte '); DROP TABLE users;--) wrote…</i></summary><blockquote>

`Factory.insert(:file, similarities)` ?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/helix/6)*

**Comment by mememori** [comment]
:lgtm:

---

Reviewed 2 of 7 files at r4, 6 of 6 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/helix/6#-:-Kfvd5_FXeI1P1y4aXK3:bnfp4nl)*

**Pull request merged**